### PR TITLE
Compatibility with SHA256 ssh keys

### PR DESCRIPTION
**Pull request by IlyaSemenov**
The current `sshcommand list` output is useless on modern Linuxes (Ubuntu 16.04 LTS in particular), and wasn't very useful even when it worked on older distros with old-style SSH fingerprints.

This PR reworks `sshcommand list` to output pairs of `username ssh-fingerprint` for each user added by `sshcommand create`.

See https://github.com/dokku/dokku/issues/2343 for the related Dokku issue (if this PR is merged, the issue should be closed).

### CODE REVIEW

**Comment by josegonzalez** [comment]
I haven't tried out your change, but this is what the current setup looks like (confirmed it has to do with the fingerprint in use):

```
# dokku ssh-keys:list
61:21:1f:88:7f:86:d4:3a:68:9f:18:aa:41:4f:bc:3d NAME="default"
 SSHCOMMAND_ALLOWED_KEYS: no-agent-forwarding,no-user-rc,no-X11-forwarding,no-port-forwarding
61:21:1f:88:7f:86:d4:3a:68:9f:18:aa:41:4f:bc:3d NAME="default"
 SSHCOMMAND_ALLOWED_KEYS: no-agent-forwarding,no-user-rc,no-X11-forwarding,no-port-forwarding
61:21:1f:88:7f:86:d4:3a:68:9f:18:aa:41:4f:bc:3d NAME="default"
 SSHCOMMAND_ALLOWED_KEYS: no-agent-forwarding,no-user-rc,no-X11-forwarding,no-port-forwarding
```

What does the new output look like?

**Comment by IlyaSemenov** [comment]
On Ubuntu 16.04, the new output looks like this:

```console
# dokku ssh-keys:list
admin SHA256:y/lCsvs2fCWjcYc0whZFYQ3UX/OO2qv+n6wE7D9dCeI
gitlab SHA256:sxDqS3wXgZ7RUfzHKLbJElMPwqEg14op/bTk/qqmiwI
```

In Debian jessie (your `Dockerfile.test` environment), the output looks like this:

```console
user1 6b:4f:f9:1a:1e:1b:7a:82:21:29:1d:21:96:6d:e5:22
```

I didn't see any real use for the extra output. Please correct me if I got it wrong.

For comparison, the output with sshcommand from the current master (9ce3fdce3b4fde1ae49a287380f5d7c920140757) looks like this on Ubuntu 16.04:

```console
# dokku ssh-keys:list
 SSHCOMMAND_ALLOWED_KEYS: no-agent-forwarding,no-user-rc,no-X11-forwarding,no-port-forwarding
 SSHCOMMAND_ALLOWED_KEYS: no-agent-forwarding,no-user-rc,no-X11-forwarding,no-port-forwarding
```

**Comment by IlyaSemenov** [comment]
I updated the PR.

Debian jessie (with custom `SSHCOMMAND_ALLOWED_KEYS`):

```
2c:c9:ce:02:d8:0c:22:26:5c:84:5e:5b:2f:86:04:a5 NAME="user1" SSHCOMMAND_ALLOWED_KEYS="keys-user1"
```

Ubuntu xenial:

```
SHA256:y/lCsvs2fCWjcYc0whZFYQ3UX/OO2qv+n6wE7D9dCeI NAME="admin" SSHCOMMAND_ALLOWED_KEYS="no-agent-forwarding,no-user-rc,no-X11-forwarding,no-port-forwarding"
SHA256:sxDqS3wXgZ7RUfzHKLbJElMPwqEg14op/bTk/qqmiwI NAME="gitlab" SSHCOMMAND_ALLOWED_KEYS="no-agent-forwarding,no-user-rc,no-X11-forwarding,no-port-forwarding"
```

As you can see I unified the syntax for printing `NAME` and `SSHCOMMAND_ALLOWED_KEYS` (why they were different?), but it's trivial to make the PR fully backwards compatible.

**Comment by josegonzalez** [comment]
@u2mejc this look good to you? I can make an sshcommand release if so.

**Comment by pascalw** [comment]
@IlyaSemenov as discussed on https://github.com/dokku/dokku/issues/2343 here's what my `authorized_keys` looks like, produced by an earlier version of `sshcommand` (not sure exactly which):

```
command="FINGERPRINT=9f:7d:fd:8e:48:86:c5:ed:41:82:f7:df:3c:2c:18:9e NAME=admin `cat /home/dokku/.sshcommand` $SSH_ORIGINAL_COMMAND",no-agent-forwarding,no-user-rc,no-X11-forwarding,no-port-forwarding ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAx6lOy3WkAieirh3UgZK0x2wwUkmaPeR+tJca4ZXGRYAh/7rqX3MHjIvgne8KDW2KGvPdy5BJpY3wkix9vRF7XU7nrl/dGnHWUT+8IcCKuAKiJQvVU1ZqYrvUXnRMeoRr1diHWoixo9BRxzu4GS4Gg+cFX3wH+qsJvMm5LollUUWy2RKWK2l3RFR6XQ3sUtvPe7FSSI+QYgst1HYQk50tigp1RYIFiLYXnvZrnAFQ05N4PjP15rTJRkOTatuDUzDB9lzEQjUq/Ew0zV8JEIn6Svp68UmeK+IWSMNsHj30k6aiZBHWYvKkrCgqAMZJdh/7nDNePqbUftdS2On9lwENpQ==
```

As you can see the main difference is the `NAME=admin` value is not quoted.

File: sshcommand [hunk 1]
```diff
@@ -172,8 +172,8 @@ sshcommand-list() {
   userhome=$(sh -c "echo ~$USER")
   [[ -e "$userhome/.ssh/authorized_keys" ]] || log-fail "authorized_keys not found for $USER"
   [[ -s "$userhome/.ssh/authorized_keys" ]] || log-fail "authorized_keys is empty for $USER"
-  grep -oE '[a-f0-9]{2}(:[a-f0-9]{2}){15}\ NAME=\\".*\\"|SSH_ORIGINAL_COMMAND\",[a-zA-Z0-9,\-]*' "$userhome/.ssh/authorized_keys" | \
-    sed -e 's/\\//g' -e 's/SSH_ORIGINAL_COMMAND",/ SSHCOMMAND_ALLOWED_KEYS: /g'
+  sed --silent --regexp-extended 's/^command="FINGERPRINT=(\S+) NAME=(\\"|)(.*)\2 `.*",(\S+).*/\1 NAME="\3" SSHCOMMAND_ALLOWED_KEYS="\4"/p' "$userhome/.ssh/authorized_key
+s"
```
Comment (josegonzalez) [inline sshcommand]: Why the extra newline?

Comment (IlyaSemenov) [inline sshcommand]: @josegonzalez My bad, mistake during copy/paste from a different server. Fixed and re-ran tests.

**Comment by josegonzalez** [comment]
Can you add unit tests for both versions of the fingerprint, as well as quoted and unquoted NAME?

**Comment by u2mejc** [comment]
@IlyaSemenov Just took the time to test this, looks pretty, nice work. Can you change the delimiters for your capture group, or add to your dangling or? (ie `|`)

Once it captures NAME, with and without quotes, and tests are happy, this will LGTM!

**Comment by IlyaSemenov** [comment]
@u2mejc I added tests for 3 variants (md5 fingerprint, sha256 fingerprint, no quotes around the key name). Please make sure you are okay with how I rearranged files in fixtures directory.

Not sure which of the capture groups did you mean? The one around the name already has dangling "or" (that's how it captures unquoted names).

**Comment by josegonzalez** [comment]
@u2mejc this look good to you?

**Pull request merged**